sky/provision/azure/instance.py

```diff
@@ -45,64 +141,412 @@ def get_azure_sdk_function(client: Any, function_name: str) -> Callable:
+    template_params['vmName'] = vm_name
+    template_params['provisionPublicIp'] = not use_internal_ips
+    template_params['vmTags'] = node_tags
+    template_params['vmCount'] = count
+    template_params['msi'] = provider_config['msi']
+    template_params['nsg'] = provider_config['nsg']
+    template_params['subnet'] = provider_config['subnet']
+    template_params['cloudInitSetupCommands'] = (base64.b64encode(
+        template_params['cloudInitSetupCommands'].encode('utf-8')).decode(
+            'utf-8'))
+
+    if node_config.get('need_nvidia_driver_extension', False):
+        # pylint: disable=line-too-long
+        # Configure driver extension for A10 GPUs. A10 GPUs requires a
+        # special type of drivers which is available at Microsoft HPC
+        # extension. Reference: https://forums.developer.nvidia.com/t/ubuntu-22-04-installation-driver-error-nvidia-a10/285195/2
+        for r in template['resources']:
+            if r['type'] == 'Microsoft.Compute/virtualMachines':
+                # Add a nested extension resource for A10 GPUs
+                r['resources'] = [
+                    {
+                        'type': 'extensions',
+                        'apiVersion': '2015-06-15',
+                        'location': '[variables(\'location\')]',
+                        'dependsOn': [
+                            '[concat(\'Microsoft.Compute/virtualMachines/\', parameters(\'vmName\'), copyIndex())]'
+                        ],
+                        'name': 'NvidiaGpuDriverLinux',
+                        'properties': {
+                            'publisher': 'Microsoft.HpcCompute',
+                            'type': 'NvidiaGpuDriverLinux',
+                            'typeHandlerVersion': '1.9',
+                            'autoUpgradeMinorVersion': True,
+                            'settings': {},
+                        },
+                    },
+                ]
+                break
+
+    parameters = {
+        'properties': {
+            'mode': azure.deployment_mode().incremental,
+            'template': template,
+            'parameters': {
+                key: {
+                    'value': value
+                } for key, value in template_params.items()
+            },
+        }
+    }
+
+    create_or_update = _get_azure_sdk_function(
+        client=resource_client.deployments, function_name='create_or_update')
+    create_or_update(
+        resource_group_name=resource_group,
+        deployment_name=vm_name,
+        parameters=parameters,
+    ).wait()
+    filters = {
+        constants.TAG_RAY_CLUSTER_NAME: cluster_name_on_cloud,
+        _TAG_SKYPILOT_VM_ID: vm_id
+    }
+    instances = _filter_instances(compute_client, resource_group, filters)
+    assert len(instances) == count, (len(instances), count)
+    return instances
+
+
+def run_instances(region: str, cluster_name_on_cloud: str,
+                  config: common.ProvisionConfig) -> common.ProvisionRecord:
```

Comment:
nit: I'm feeling like this function is too large to read. Maybe leave a TODO to separate this into multiple smaller functions?